File: android/rxflux-core/src/main/java/com/klefmusic/rxfluxcore/Store.kt
package com.klefmusic.rxfluxcore

import com.jakewharton.rx3.replayingShare
import com.jakewharton.rxrelay3.PublishRelay
import com.klefmusic.rxfluxcore.models.Action
import io.reactivex.rxjava3.annotations.NonNull
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single


interface ReduxStore<S> {

    fun dispatch(action: Action)

    val actions: Observable<Action>

    fun currentViewState(): S

    val updates: Observable<S>

}

/**
 * Simplest possible redux implementation
 */
fun <S> Observable<Action>.reduxStore(
    initialStateSupplier: Single<S>,
    reducer: Reducer<S>
): @NonNull Observable<S> =
    initialStateSupplier
        .flatMapObservable { initialState ->
            scan(initialState) { oldState, result ->
                reducer(oldState, result)
            }
        }
        .distinctUntilChanged()
        .replayingShare()

fun <S> createStore(
    initialStateSupplier: Single<S>,
    reducer: Reducer<S>
): ReduxStore<S> = object : ReduxStore<S> {

    private val resultsRelay = PublishRelay.create<Action>()

    override val actions: Observable<Action> = resultsRelay.hide()

    override fun dispatch(action: Action) = resultsRelay.accept(action)

    override fun currentViewState(): S = updates.blockingFirst()

    override val updates: Observable<S> = actions.reduxStore(initialStateSupplier, reducer)
}

fun <S> ReduxStore<S>.withMiddleware(middleware: (store: ReduxStore<S>) -> Completable): ReduxStore<S> {
    val wrapped = this
    val middlewareCompletable = middleware(wrapped)
    return object : ReduxStore<S> {

        override val actions: Observable<Action> = wrapped.actions

        override fun dispatch(action: Action) = wrapped.dispatch(action)

        override val updates: Observable<S> =
            Observable.merge(
                wrapped.updates,
                middlewareCompletable.toObservable()
            )

        override fun currentViewState(): S = wrapped.currentViewState()
    }
}

File: iOS/mobileflux/mobileflux/Store.swift
//
//  Store.swift
//  mobileflux
//
//  Created by RORY KELLY on 23/09/2020.
//

import Foundation
import RxSwift
import RxCocoa

protocol ReduxStore {
    associatedtype S : State

    func dispatch(action: Action)

    var actions: Observable<Action> { get }

    func currentViewState() -> S

    var updates: Observable<S>{ get }

}


extension Observable where Observable.Element == Action {
   
    func reduxStore<VS: State>(
        initialStateSupplier: Single<VS>,
        reducer: @escaping Reducer<VS>
    )-> Observable<VS> {
        return initialStateSupplier
            .asObservable()
            .flatMap { initialState -> Observable<VS> in
                return self.scan(initialState) { oldState, result in
                    reducer(oldState, result)
                }
            }
            .distinctUntilChanged()
            .share(replay: 1, scope: .whileConnected)
    }
}

typealias Middleware<S: State> = (_ actions: Observable<Action>, _ dispatch: @escaping (Action) -> Void, _ state: () -> S ) -> Observable<S>


open class Store<S: State>: ReduxStore {
    typealias S = S
    
    private let resultsRelay = PublishRelay<Action>()
    var actions: Observable<Action>
    var updates: Observable<S>
 
    init(
        initialStateSupplier: Single<S>,
        reducer: @escaping Reducer<S>,
        middleware: Middleware<S>
    ) {
        actions = resultsRelay.asObservable()
        updates = actions.reduxStore(initialStateSupplier: initialStateSupplier, reducer: reducer)
        updates = Observable.merge(updates, middleware(actions, dispatch, currentViewState).asObservable())
    }
    
    func dispatch(action: Action) {
        resultsRelay.accept(action)
    }
    
    func currentViewState() -> S {
        return updates.first() as! S
    }
    
}

